### PR TITLE
Improve variability of `$rules` argument in `Validator::validate()`

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #292

### CODE REVIEW

File: src/Rule/Nested.php [hunk 1]
```diff
@@ -132,18 +135,13 @@ private function prepareRules(iterable|object|string|null $source): ?iterable
 
         if ($source instanceof RulesProviderInterface) {
             $rules = $source->getRules();
-            return $this->normalizeRules ? $this->normalizeRules($rules) : $rules;
```
Comment (vjik) [inline src/Rule/Nested.php:135]: Why we don't can make return here?

File: src/Rule/Nested.php [hunk 2]
```diff
@@ -132,18 +135,13 @@ private function prepareRules(iterable|object|string|null $source): ?iterable
...
+        } elseif (!$source instanceof Traversable && !is_array($source)) {
+            $rules = (new AttributesRulesProvider($source, $this->rulesPropertyVisibility))->getRules();
```
Comment (vjik) [inline src/Rule/Nested.php:139]: Why we don't can make return here?

File: src/Validator.php [hunk 3]
```diff
@@ -46,13 +49,27 @@ public function __construct(
...
+     * @param class-string|iterable<Closure|Closure[]|RuleInterface|RuleInterface[]>|RulesProviderInterface|null $rules
+     * @param int $rulesPropertyVisibility What visibility levels to use when reading rules from the class specified in
+     * `$rules` argument.
+     *
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
      */
-    public function validate(mixed $data, ?iterable $rules = null): Result
-    {
+    public function validate(
```
Comment (samdark) [inline src/Validator.php:59]: It's not the same as in the interface: https://github.com/yiisoft/validator/blob/master/src/ValidatorInterface.php#L10

File: src/Rule/Nested.php (hunk 1 above)
Comment (arogachev) [reply src/Rule/Nested.php:135]: To perform check and normalization for all cases and prevent duplication of corresponding code. Or can we guarantee that in other cases it's not needed? Maybe normalization can be skipped in certain scenarios, but what about checking? I suggest to leave it like that.

File: src/Rule/Nested.php (hunk 2 above)
Comment (arogachev) [reply src/Rule/Nested.php:139]: https://github.com/yiisoft/validator/pull/303#discussion_r965789351

File: src/Validator.php (hunk 3 above)
Comment (arogachev) [reply src/Validator.php:59]: Good catch! :+1: Then it's better to move it to a property like in `Nested` rule.

**Review by samdark — APPROVED** [review]

**Pull request merged**